File: othello_python/othello/player.py
"""
Class Player
Defines one player for Othello
Akseli Lukkarila
2019-2023
"""
import random
import time
from typing import Self

from board import Board
from colorprint import Color, print_color, print_error
from utils import Disk, Move, Square


class Player:
    """Defines one player that can be either human or computer controlled."""

    def __init__(self, disk: Disk, human=True, show_helpers=True):
        self.can_play: bool = True
        self._disk: Disk = disk
        self._human: bool = human
        self._rounds_played: int = 0
        self._show_helpers: bool = show_helpers

    @classmethod
    def black(cls) -> Self:
        """Shorthand to initialize a new player for black disks."""
        return Player(Disk.BLACK)

    @classmethod
    def white(cls) -> Self:
        """Shorthand to initialize a new player for white disks."""
        return Player(Disk.WHITE)

    def play_one_move(self, board: Board) -> None:
        """Play one round as this player."""
        print(f"Turn: {str(self._disk)}")
        moves = board.possible_moves(self._disk)
        if moves:
            self.can_play = True
            if self._human and self._show_helpers:
                board.print_moves(moves)

            chosen_move = (
                self._get_human_move(moves) if self._human else self._get_computer_move(moves)
            )
            board.place_disk(chosen_move)
            board.print_score()
            self._rounds_played += 1
            time.sleep(1)
        else:
            self.can_play = False
            print_color("  No moves available...", Color.yellow)

    def reset(self) -> None:
        """Reset player status for a new game."""
        self.can_play = True
        self._rounds_played = 0

    def set_human(self, is_human: bool) -> None:
        """Set the player as human or computer controlled."""
        self._human = is_human

    @staticmethod
    def _get_computer_move(moves: list[Move]) -> Move:
        """Return move chosen by computer."""
        print("  Computer plays...")
        # Wait a bit and pick a random move
        time.sleep(random.uniform(1.0, 2.0))
        chosen_move = random.choice(moves)
        print(f"  {chosen_move.square} -> {chosen_move.value}")
        return chosen_move

    def _get_human_move(self, moves: list[Move]) -> Move:
        """Return move chosen by a human player."""
        while True:
            square = self._get_square()
            # Check that the chosen square is actually one of the possible moves
            move = next((move for move in moves if move.square == square), None)
            if move:
                return move

            print_error(f"  Can't place a {self._disk} disk in square {square}!")

    @staticmethod
    def _get_square() -> Square:
        """Ask human player for square coordinates."""
        while True:
            try:
                pos = input("  Give disk position (x,y): ")
                x, y = (int(i) for i in pos.split(",") if i.strip())
                return Square(x, y)
            except ValueError:
                print_error("  Give coordinates in the form 'x,y'!")

    def _type_string(self) -> str:
        """Return player type description string."""
        return "Human   " if self._human else "Computer"

    def __str__(self) -> str:
        return f"{str(self._disk)} | {self._type_string()} | Moves: {self._rounds_played}"

File: othello_rust/src/player.rs
//==========================================================
// Player
// Defines one player for Othello
// Akseli Lukkarila
// 2019-2023
//==========================================================

extern crate colored;

use colored::Colorize;
use rand::{self, Rng};

use std::io::Write;
use std::{fmt, io};
use std::{thread, time::Duration};

use crate::board::Board;
use crate::colorprint::print_error;
use crate::utils::{Disk, Move, Square};

/// Defines one player that can be either human or computer controlled.
#[derive(Debug)]
pub(crate) struct Player {
    pub can_play: bool,
    disk: Disk,
    human: bool,
    rounds_played: u32,
    show_helpers: bool,
}

impl Player {
    /// Initialize new player for the given disk color.
    pub fn new(disk: Disk) -> Player {
        Player {
            disk,
            human: true,
            can_play: true,
            show_helpers: true,
            rounds_played: 0,
        }
    }

    /// Shorthand to initialize a new player for black disks.
    pub fn black() -> Player {
        Player::new(Disk::Black)
    }

    /// Shorthand to initialize a new player for white disks.
    pub fn white() -> Player {
        Player::new(Disk::White)
    }

    /// Play one round as this player.
    pub fn play_one_move(&mut self, board: &mut Board) {
        println!("Turn: {}", self.disk);
        let moves = board.possible_moves(self.disk);
        if !moves.is_empty() {
            self.can_play = true;
            if self.human && self.show_helpers {
                board.print_moves(&moves)
            }
            let chosen_move = if self.human {
                self.get_human_move(&moves)
            } else {
                self.get_computer_move(&moves)
            };
            board.place_disk(&chosen_move);
            board.print_score();
            self.rounds_played += 1;
            thread::sleep(Duration::from_secs(1));
        } else {
            self.can_play = false;
            println!("{}", "  No moves available...".yellow());
        }
    }

    /// Reset player status for a new game.
    pub fn reset(&mut self) {
        self.can_play = true;
        self.rounds_played = 0;
    }

    /// Set the player as human or computer controlled.
    pub fn set_human(&mut self, is_human: bool) {
        self.human = is_human;
    }

    /// Return move chosen by computer.
    fn get_computer_move(&self, moves: &Vec<Move>) -> Move {
        println!("  Computer plays...");
        // Wait a bit and pick a random move
        thread::sleep(Duration::from_millis(
            rand::thread_rng().gen_range(1000..2000),
        ));
        let chosen_move = moves[rand::thread_rng().gen_range(0..moves.len())].clone();
        println!("  -> {}", chosen_move.square);
        chosen_move
    }

    /// Return move chosen by a human player.
    fn get_human_move(&self, moves: &[Move]) -> Move {
        loop {
            let square = Self::get_square();
            // Check that the chosen square is actually one of the possible moves
            if let Some(valid_move) = moves.iter().find(|m| m.square == square) {
                return valid_move.clone();
            }
            print_error(&format!(
                "  Can't place a {} disk in square {}!",
                self.disk, square
            ));
        }
    }

    /// Ask human player for square coordinates.
    fn get_square() -> Square {
        loop {
            print!("  Give disk position (x,y): ");
            let mut input = String::new();
            io::stdout().flush().unwrap();
            io::stdin().read_line(&mut input).expect("Input failed");
            let values: Vec<&str> = input.trim().split(',').collect();
            // TODO: nicer error handling
            if values.len() == 2 {
                let x: isize = values[0].parse().unwrap_or(-1);
                let y: isize = values[1].parse().unwrap_or(-1);
                if x >= 0 && y >= 0 {
                    return Square { x, y };
                }
            }
            print_error("  Give coordinates in the form 'x,y'!")
        }
    }
}

impl fmt::Display for Player {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{} | {} | Moves: {}",
            self.disk,
            if self.human { "Human   " } else { "Computer" },
            self.rounds_played
        )
    }
}
